src/walker.ts

```diff
@@ -0,0 +1,62 @@
+import { PotentialStep, Pipeline, Conditional, Command, CommandStep } from '.';
+import { getAndCacheDependency } from './conditional';
+
+export type Entity = Pipeline | PotentialStep | Command;
+
+async function walkEntity<T extends Entity>(
+  entity: T,
+  nodeCache: Map<any, any>,
+  mutator?: (entity: Entity) => any,
+): Promise<T> {
+  if (nodeCache.has(entity)) {
+    return nodeCache.get(entity);
+  }
+
+  const cache = new Map();
+
+  if (entity instanceof Conditional) {
+    if (await entity.accept()) {
+      entity = <any>await getAndCacheDependency(cache, entity);
+    } else {
+      return entity;
+    }
+  }
+
+  if (entity instanceof Pipeline) {
+    const steps = entity.steps;
```

Comment:
I think mutating the steps during the iteration makes it so the mutator can see half mutated step. It's safer to create a new list of steps, and replace the old steps with the new steps at the end of the iteration.

Reply:
We will have to modify all these attributes so they are mutable, which is probably ok, since that's the intention of this

Reply:
Also, is this an issue considering we are always processing dependencies first?

Reply:
It might be if an implementation depends on it, after all since we're modifying the entity that is being passed in, the calling code will have a reference to that entity. It would be different if we were creating a new copy of the entity and returning the modified copy without modifying the entity that is passed in.

Reply:
actioned this